create_kubeconfig.go

```diff
@@ -0,0 +1,102 @@
+package kubeconfig
+
+import (
+	"github.com/giantswarm/microerror"
+	yaml "gopkg.in/yaml.v2"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+// Kubeconfig is a struct used to create a kubectl configuration YAML file
+type KubeConfigValue struct {
+	APIVersion     string                   `yaml:"apiVersion"`
+	Kind           string                   `yaml:"kind"`
+	Clusters       []KubeconfigNamedCluster `yaml:"clusters"`
+	Users          []KubeconfigUser         `yaml:"users"`
+	Contexts       []KubeconfigNamedContext `yaml:"contexts"`
+	CurrentContext string                   `yaml:"current-context"`
+	Preferences    struct{}                 `yaml:"preferences"`
+}
+
+// KubeconfigUser is a struct used to create a kubectl configuration YAML file
+type KubeconfigUser struct {
+	Name string                `yaml:"name"`
+	User KubeconfigUserKeyPair `yaml:"user"`
+}
+
+// KubeconfigUserKeyPair is a struct used to create a kubectl configuration YAML file
+type KubeconfigUserKeyPair struct {
+	ClientCertificateData string `yaml:"client-certificate"`
+	ClientKeyData         string `yaml:"client-key"`
+}
+
+// KubeconfigNamedCluster is a struct used to create a kubectl configuration YAML file
+type KubeconfigNamedCluster struct {
+	Name    string            `yaml:"name"`
+	Cluster KubeconfigCluster `yaml:"cluster"`
+}
+
+// KubeconfigCluster is a struct used to create a kubectl configuration YAML file
+type KubeconfigCluster struct {
+	Server                   string `yaml:"server"`
+	CertificateAuthorityData string `yaml:"certificate-authority-data"`
+}
+
+// KubeconfigNamedContext is a struct used to create a kubectl configuration YAML file
+type KubeconfigNamedContext struct {
+	Name    string            `yaml:"name"`
+	Context KubeconfigContext `yaml:"context"`
+}
+
+// KubeconfigContext is a struct used to create a kubectl configuration YAML file
+type KubeconfigContext struct {
+	Cluster string `yaml:"cluster"`
+	User    string `yaml:"user"`
+}
+
+func Marshal(config *KubeConfigValue) ([]byte, error) {
+	bytes, err := yaml.Marshal(config)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	return bytes, nil
+}
+
+func Unmarshal(bytes []byte) (*KubeConfigValue, error) {
+	var kubeConfig KubeConfigValue
+	err := yaml.Unmarshal(bytes, &kubeConfig)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	return &kubeConfig, nil
+}
+
+func (k *KubeConfig) NewKubeConfigForRESTConfig(config *rest.Config) ([]byte, error) {
+	kubeConfig := KubeConfigValue{
+		Clusters: []KubeconfigNamedCluster{
+			{
+				Cluster: KubeconfigCluster{
+					Server: config.Host,
```

Comment:
I'm not sure what I can add up more than the host, since `rest.Config` do not have many relevant variables with kubeConfig, can someone help me to understand this?

Reply:
In the kubeconfig you need the context and user as well as the cluster which has the API server hostname.

As an example this is a kubeconfig generated by gsctl. This is using file paths for the TLS certificates but we should also support binary data. 

```yaml
apiVersion: v1
clusters:
- cluster:
    certificate-authority: /Users/ross/.config/gsctl/certs/3dyqu-ca.crt
    server: https://api.3dyqu.k8s.ginger.eu-central-1.aws.gigantic.io
  name: giantswarm-3dyqu
contexts:
- context:
    cluster: giantswarm-3dyqu
    namespace: giantswarm
    user: giantswarm-3dyqu-user
  name: giantswarm-3dyqu
users:
- name: giantswarm-3dyqu-user
  user:
    client-certificate: /Users/ross/.config/gsctl/certs/3dyqu-3393bb65ea-client.crt
    client-key: /Users/ross/.config/gsctl/certs/3dyqu-3393bb65ea-client.key
```

Reply:
Yes, For sure I knew that structure of kubeconfig but unfortunately finding out matching client-certificate, authority, namespace, all of them in rest.Config structure is that I don't know how to find out correct values inside of it, do you know that any code that take similar action?

Reply:
There is some similar code in operatorkit. Maybe it helps?

https://github.com/giantswarm/operatorkit/blob/9b85f61ceb90ed074402e1cb1ba005e4170ece9e/client/k8srestconfig/k8s_rest_config.go#L137-L148

Reply:
So below is the `rest.Config` file, from my kubeconfig files. 
```
{Host:                "https://192.168.64.35:8443",
 APIPath:             "",
 ContentConfig:       {AcceptContentTypes:   "",
                       ContentType:          "",
                       GroupVersion:         nil,
                       NegotiatedSerializer: nil},
 Username:            "",
 Password:            "",
 BearerToken:         "",
 BearerTokenFile:     "",
 Impersonate:         {UserName: "",
                       Groups:   [],
                       Extra:    {}},
 AuthProvider:        nil,
 AuthConfigPersister: nil,
 ExecProvider:        nil,
 TLSClientConfig:     {Insecure:   false,
                       ServerName: "",
                       CertFile:   "/Users/jihyukbok/.minikube/client.crt",
                       KeyFile:    "/Users/jihyukbok/.minikube/client.key",
                       CAFile:     "/Users/jihyukbok/.minikube/ca.crt",
                       CertData:   [],
                       KeyData:    [],
                       CAData:     []},
 UserAgent:           "",
 Transport:           nil,
 WrapTransport:       (func(http.RoundTripper) http.RoundTripper)(nil),
 QPS:                 0,
 Burst:               0,
 RateLimiter:         nil,
 Timeout:             0,
 Dial:                (func(context.Context, string, string) (net.Conn, error))(nil)}
```
As you see, it's hard to get what namespace, cluster, username we should use. 
From what I see, it's almost impossible to convert rest.Config into kubeconfig. 
Should we cover this feature on this library?

Reply:
OK yes that isn't going to work. We need to change the approach. Looking at it now.